cpp/terrain/terrain_object.cpp

```diff
@@ -26,26 +26,28 @@ TerrainObject::TerrainObject(Unit &u, bool collisions)
 	draw{[]() {}},
 	placed{false},
 	terrain{nullptr},
-	occupied_chunk_count{0},
-	parent{nullptr} {
-
-	// delete any exisitng location	
-	if (unit.location) {
-		delete unit.location;
-	}
-	unit.location = this;
+	occupied_chunk_count{0} {
 }
 
 TerrainObject::~TerrainObject() {
 	// remove all connections from terrain
+	this->unit.log(MSG(dbg) << "Cleanup terrain object");
 	this->remove();
-	for (auto &c : children) {
-		c->remove();
-		delete c;
-	}
 	this->unit.location = nullptr;
 }
 
+void TerrainObject::initialise() {
+
+	// remove any exisitng location	
+	if (unit.location) {
+		unit.location->remove();
+	}
+
+	// shared_from_this cannot be used in the constructor,
+	// so an initalise function is required
+	unit.location = this->shared_from_this();
```

Comment:
`u->location` is the single owner of the TerrainObject, right?

Which cases are there where a `TerrainObject` is constructed, and then only later assigned to the location?